web3/contract.py

```diff
@@ -117,6 +119,19 @@ def __iter__(self):
         for func in self._functions:
             yield func['name']
 
+    def __getattr__(self, function_name):
+        if '_functions' not in self.__dict__:
+            raise NoABIFunctionsFound(
+                "The abi for this contract contains no function definitions. ",
+                "Are you sure you provided the correct contract abi?"
+            )
+        elif function_name not in self.__dict__['_functions']:
+            raise MismatchedABI(
+                "The function '{}' was not found in this contract's abi. ".format(function_name),
+                "Are you sure you provided the correct contract abi?"
+            )
+        return getattr(self, function_name)
```

Comment:
This seems like it should be a `super()` call.

Reply:
Actually it should raise `AttributeError`. `__getattr__` is only called when the normal attribute lookup fails. If we reach this point the lookup has already failed, we are just deciding which error message to show. I'll push an update with this behaviour.

Reply:
My thought was that the `super()` would indeed raise an `AttributeError` but that we shouldn't presume that there isn't additional `__getitem__` logic to be executed.

Reply:
Makes sense. I didn't add it initially because both `__init__` and `__getitem__` already presume this is a top-level class and do not call `super()`.

I'll need to implement it as `return super().__getattribute__(function_name)` - because it is a top-level class `__getattr__` is not exposed through super calls. If `super().__getattribute__` can't find the attribute it will fall back to `super().__getattr__` if it is defined, so the effect is identical.